packages/amplify_auth_cognito/example/ios/amplify_auth_cognito_tests/MockCognito.swift

```diff
@@ -0,0 +1,224 @@
+/*
+* Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+*
+* Licensed under the Apache License, Version 2.0 (the "License").
+* You may not use this file except in compliance with the License.
+* A copy of the License is located at
+*
+*  http://aws.amazon.com/apache2.0
+*
+* or in the "license" file accompanying this file. This file is distributed
+* on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+* express or implied. See the License for the specific language governing
+* permissions and limitations under the License.
+*/
+
+import Amplify
+import AmplifyPlugins
+import AWSCore
+import AWSPluginsCore
+import AWSMobileClient
+@testable import amplify_auth_cognito
+
+class MockCognito: AuthCognitoProtocol {
+    
+    var returnError: Bool = false
+    
+    func onSignUp(flutterResult: @escaping FlutterResult, request: FlutterSignUpRequest) {
+        if (!returnError) {
+            var signUpRes = Result<AuthSignUpResult,AuthError>.success(
+                AuthSignUpResult(AuthSignUpStep.confirmUser(AuthCodeDeliveryDetails(destination: DeliveryDestination.email(_email)), ["foo": "bar"])))
+            if (((request.options!["userAttributes"] as! Dictionary<String, Any>)["phone_number"]) as? String == _phoneNumber) {
+                signUpRes = Result<AuthSignUpResult,AuthError>.success(
+                AuthSignUpResult(AuthSignUpStep.confirmUser(AuthCodeDeliveryDetails(destination: DeliveryDestination.phone(_phoneNumber)), ["foo": "bar"])))
+            } else if (((request.options!["userAttributes"] as! Dictionary<String, Any>)["sms"]) as? String == _phoneNumber) {
+                signUpRes = Result<AuthSignUpResult,AuthError>.success(
+                AuthSignUpResult(AuthSignUpStep.confirmUser(AuthCodeDeliveryDetails(destination: DeliveryDestination.sms(_phoneNumber)), ["foo": "bar"])))
+            } else if (((request.options!["userAttributes"] as! Dictionary<String, Any>)["custom"]) as? String == _phoneNumber) {
+                signUpRes = Result<AuthSignUpResult,AuthError>.success(
+                AuthSignUpResult(AuthSignUpStep.confirmUser(AuthCodeDeliveryDetails(destination: DeliveryDestination.unknown(_phoneNumber)), ["foo": "bar"])))
+            } else if (request.options!["complete"] as? Bool == true) {
+                signUpRes = Result<AuthSignUpResult,AuthError>.success(AuthSignUpResult(AuthSignUpStep.done))
+            }
+            let signUpData = FlutterSignUpResult(res: signUpRes)
+            flutterResult(signUpData)
+        } else {
+            let authError = AuthError.service("Invalid username", "Enter a username", nil)
+            handleAuthError(error: authError, flutterResult: flutterResult, msg: FlutterAuthErrorMessage.SIGNUP.rawValue)
+        }
```

Comment:
I feel there might be better way to decouple this Mock and tests so that tests can **define** how the mock should behave for this particular unit test execution. This will also enable more unhappy test scenarios I believe. We can sync up on this offline.

Reply:
Done, as discussed this morning.